Modules/_testcapi/heaptype.c

```diff
@@ -973,6 +973,110 @@ static PyType_Spec HeapCTypeSetattr_spec = {
     HeapCTypeSetattr_slots
 };
 
+PyDoc_STRVAR(HeapCCollection_doc,
+"Tuple-like heap type that uses PyObject_GetItemData for items.");
+
+static PyObject*
+HeapCCollection_new(PyTypeObject *subtype, PyObject *args, PyObject *kwds)
+{
+    PyObject *self = NULL;
+    PyObject *result = NULL;
+
+    Py_ssize_t size = PyTuple_GET_SIZE(args);
+    self = subtype->tp_alloc(subtype, size);
+    if (!self) {
+        goto finally;
+    }
+    PyObject **data = PyObject_GetItemData(self);
+    if (!data) {
+        goto finally;
+    }
```

Comment:
You need to set an exception before erring out.

Reply:
`PyObject_GetItemData` sets an exception on error.

Reply:
Like most other C API lunctions, whenever `PyObject_GetItemData` returns NULL, it sets an exception. 
If it *currently* has no error cases, and won't actually return `NULL`, that's an internal implementation detail. The tests show how a user should use the API: you should check for errors and propagate exceptions.